### ISSUE TITLE
Why use the block diffusion by default?

### ISSUE THREAD

**Issue by SeaBird-Go** [hustvl/DiffusionVL#3]
Thank you for sharing this wonderful work. After reading your paper, I'm curious about some aspects to ask you to clarify:
1. LLaDA and the latest multi-modal DLMs like MMaDA utilize the standard masking generative modeling during training by default, in which they **randomly** mask the **whole** GT tokens to supervise the predicted tokens with CE-loss. Why did you choose to use the block diffusion paradigm in DiffusionVL? Is it possible to translate the existing AR VLMs into the standard DLMs?

2. In Figure 4, you compare the DiffusionVL with LLaDA-V to validate the inference speed strength of the DiffusionVL. How about the speed analysis with the corresponding AR-VLMs under different block size settings?

**Comment by xiazhi1** [comment]
> Thank you for sharing this wonderful work. After reading your paper, I'm curious about some aspects to ask you to clarify:
> 
> 1. LLaDA and the latest multi-modal DLMs like MMaDA utilize the standard masking generative modeling during training by default, in which they **randomly** mask the **whole** GT tokens to supervise the predicted tokens with CE-loss. Why did you choose to use the block diffusion paradigm in DiffusionVL? Is it possible to translate the existing AR VLMs into the standard DLMs?
> 2. In Figure 4, you compare the DiffusionVL with LLaDA-V to validate the inference speed strength of the DiffusionVL. How about the speed analysis with the corresponding AR-VLMs under different block size settings?

@SeaBird-Go  Hi, thank you for your interest in our work and for these insightful questions! Here are the clarifications regarding your points:

**Regarding the choice of Block Diffusion:**

We chose the block diffusion paradigm because, compared to the full diffusion paradigm (as seen in LLaDA and MMaDA), it supports more flexible generation and enables highly efficient KV-cache reuse. This allows us to achieve extreme inference speeds, as demonstrated in Figure 4 of our paper.

Conceptually, block diffusion acts as an interpolation between the Auto-Regressive (AR) and full diffusion paradigms. If the block size is set to the full sequence length, it effectively behaves the same as full diffusion. Therefore, we believe block diffusion unifies the existing full diffusion paradigm.

As shown in Table 4, a larger block size (implying a larger deviation from the AR paradigm) leads to a slight decrease in performance. In our early experiments, we also found that directly translating from the AR paradigm to full diffusion yielded worse results compared to the block diffusion approach.

**Regarding speed comparison with AR-VLMs:**

We have indeed compared it with AR-VLMs internally. However, current support for inference engines and Flash Attention optimizations for this architecture is not yet fully mature.

In our current tests with a block-size=8 and fewer denoising steps, DiffusionVL allows for faster inference speeds than Qwen2.5-VL (which utilizes Flash Attention), though there is still a performance gap. Narrowing this gap while maintaining speed advantages is a key focus for our future optimization.

**Comment by SeaBird-Go** [comment]
Thanks for explaining my confusion. 

**Why I have concerns about the block diffusion:**

Although the block diffusion is a flexible strategy between AR and DLMs. When using the block diffusion, it makes the full attention paradigm in DLMs to be a semi-autoregressive paradigm. It maybe doesn't fully exploit the potential of the DLMs. The purpose of your paper is to transfer the pre-trained AR-VLMs to DLMs, the block diffusion tend to yield better results since they share similar paradigms.

BTW, when using block diffusion, the sequence length will be almost doubled, have you adapted some approaches to reduce the GPU memory consumption?

**Confusion about the attention mask in Figure 3**:
Why the order of the $x_t^{1}$ to $x_0^{3}$ in the row direction and column direction is different in the Figure 3? I guess they should be the same like the one in the paper of block diffusion?

<img width="922" height="577" alt="Image" src="https://github.com/user-attachments/assets/3c616b09-773b-4e9b-87db-2296c02770af" />

**Comment by xiazhi1** [comment]
> Thanks for explaining my confusion.
> 
> **Why I have concerns about the block diffusion:**
> 
> Although the block diffusion is a flexible strategy between AR and DLMs. When using the block diffusion, it makes the full attention paradigm in DLMs to be a semi-autoregressive paradigm. It maybe doesn't fully exploit the potential of the DLMs. The purpose of your paper is to transfer the pre-trained AR-VLMs to DLMs, the block diffusion tend to yield better results since they share similar paradigms.
> 
> BTW, when using block diffusion, the sequence length will be almost doubled, have you adapted some approaches to reduce the GPU memory consumption?
> 
> **Confusion about the attention mask in Figure 3**: Why the order of the x t 1 to x 0 3 in the row direction and column direction is different in the Figure 3? I guess they should be the same like the one in the paper of block diffusion?
> 
> <img alt="Image" width="922" height="577" src="https://private-user-images.githubusercontent.com/117327481/536018935-3c616b09-773b-4e9b-87db-2296c02770af.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg3Mjk2ODQsIm5iZiI6MTc2ODcyOTM4NCwicGF0aCI6Ii8xMTczMjc0ODEvNTM2MDE4OTM1LTNjNjE2YjA5LTc3M2ItNGU5Yi04N2RiLTIyOTZjMDI3NzBhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDExOFQwOTQzMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MmFkZDdkOTE2ZTQwOWZiZmI3MDAyZGZjMDRjY2U5OGM3YjU5MGMxOGRjMjVjZWM4Zjc4NWYyYjllMzI2YjQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.my2Y-thns-hre6GPiJ0UryXEQMH1_Z1VpXKE5BpET4w">

@SeaBird-Go Hi, thanks for your questions. Regarding your concerns:

Regarding memory consumption: We have not yet adapted specific approaches to optimize the doubled sequence length in block diffusion. We agree that this is indeed a factor limiting GPU efficiency in the current implementation.

Regarding Figure 3: The row direction represents which positions in the column direction the current block can attend to. In Figure 3, this means that the tokens from $x_t^1$ to $x_0^3$ (in the row direction) are allowed to see their own block as well as the corresponding preceding clean blocks.
